common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -1172,47 +1634,58 @@ private void finalizePartition() throws IOException {
       // Get rid of any partial block data at the end of the file. This could either
       // be due to failure, or a request still being processed when the shuffle
       // merge gets finalized, or any exceptions while updating index/meta files.
-      dataChannel.truncate(lastChunkOffset);
+      logger.trace("{}_{} shuffleId {} shuffleMergeId {} reduceId {} "
+         + "truncating files data {} index {} meta {}", appShufflePartitionId.appId,
+          appShufflePartitionId.attemptId, appShufflePartitionId.shuffleId,
+          appShufflePartitionId.shuffleMergeId, appShufflePartitionId.reduceId,
+          lastChunkOffset, indexFile.getPos(), metaFile.getPos());
+      dataFile.getChannel().truncate(lastChunkOffset);
       indexFile.getChannel().truncate(indexFile.getPos());
       metaFile.getChannel().truncate(metaFile.getPos());
     }
 
     void closeAllFilesAndDeleteIfNeeded(boolean delete) {
       try {
-        if (dataChannel.isOpen()) {
-          dataChannel.close();
-          if (delete) {
-            dataFile.delete();
-          }
+        dataFile.close();
+        if (delete) {
+          dataFile.delete();
```

Comment:
There is a change in behavior here - we are deleting even if file was not open.
If we are simplifying to only tracking finalized merges, we remove the need for this change as well and preserve earlier behavior.